### ISSUE TITLE
BooleanProperty does not enforce boolean values

### ISSUE THREAD

**Issue by dessant** [kivy/kivy#4163]
Currently it behaves exactly like ObjectProperty, the same check is performed.

**Comment by crishoj** [comment]
Two possible approaches:

1. Raise a `ValueError` the assigned value is not a `bool`. This would be consistent with the behaviour of other properties.
2. Cast the given value to `bool` before assigning. This aligns with Python's [truth value testing](https://docs.python.org/2/library/stdtypes.html#truth-value-testing).

**Comment by pythonic64** [comment]
In favor of **approach 1**.

**Comment by matham** [comment]
Why is this an issue? In Python it's not about "truth" but about truthy. So
any value should be able to be assigned to a bool prop. The only difference
between bool and a object prop is that a bool prop indicates we just care
if it's true or not.

So any value should be assignable to a bool prop. As for converting to
bool, why? What's the point? Unless you care explicitly if it's True or
False, which should happen rarely, there's no reason for this.

**Comment by dessant** [comment]
@matham, i don't see the reason of booleanproperty's existence if a boolean value is not enforced.
ObjectProperty already serves the use case where any object must be assignable, an truthiness can be checked the same way.

**Comment by inclement** [comment]
I agree about truthiness (as discussed earlier on irc), though I do think it's not ideal to have the surprising behaviour that switching between different objects with the same truthiness will dispatch events. Perhaps the property shouldn't modify its input, but should refrain from event dispatching unless the truthiness changes (i.e. check if bool(new_object) == bool(old_object) rather than direct equality)?

**Comment by kived** [comment]
Python in general doesn't enforce types. I think if we're going to have typed properties, they should enforce it strictly, otherwise why have BooleanProperty or StringProperty? Though I would tend to prefer coercion rather than rejection.

**Comment by pythonic64** [comment]
It makes more sense to me that property defined as BooleanProperty accept only True of False as values.

**Comment by matham** [comment]
I agree with inclement about this, which is easy to implement. To me bool
prop is a statement of intent. That we just care about its truthiness. But
converting things to true/false, I wonder if it'd cause issues for e.g. NP
arrays.  Although I doubt that. So what kived suggested is fine as well.

**Comment by matham** [comment]
Rejection is definitely wrong because anything that's convertible to
true/false should be accepted. Since that's how Python works typically.
On Apr 9, 2016 4:34 PM, "Matthew Einhorn" <moiein2000@gmail.com> wrote:

I agree with inclement about this, which is easy to implement. To me bool
prop is a statement of intent. That we just care about its truthiness. But
converting things to true/false, I wonder if it'd cause issues for e.g. NP
arrays.  Although I doubt that. So what kived suggested is fine as well.

**Comment by dessant** [comment]
I see ListProperty converts a string to a list, so we have precedent for option 2 (casting the value to bool).

I would agree with option 2 if we would also make the StringProperty behavior consistent, currently it only accepts strings.

**Comment by pythonic64** [comment]
How about casting value to bool for BooleanProperty, making StringProperty consistent and adding keyword argument 'strict' to constructor, which if set to True will make property enforce strict types of values?

**Comment by dessant** [comment]
@pythonic64, I find that interesting to explore.

I find both behaviors perfectly legitimate. Personally I've never even dared to use something other than a boolean as a BooleanProperty value, nor have I seen it being done in the kivy source or third-party widgets, and I can understand why users are expecting type enforcement.

Properties currently use one or the other in a surprising/unexpected manner, sticking to one behaviour across properties by default while also having an alternative behavior seems to be the most sensible.

**Comment by matham** [comment]
I do not agree. When do you find you need a strict bool/list/whatever? In python things tend to be fluid, kinda like duck typing. So I'm not sure why you'd need a prop to enforce and be strict. If it acts like a bool and quacks like a bool it's a bool. That why we write `if val:...` or `if not val:...` and not `if val is True/False:...`.

For strings, I can see where it may be useful because of the bytes/string thing. So it may make sense to have modes, or even a `StringProperty`, `UnicodeProperty`, `BytesProperty`, and a `StringTypeProperty` (or have it as a option of `StringProperty`). But for a bool? As a option I don't think it makes sense, rather we should choose how it should act for the property. Do you ever find yourself wishing you could make it strict/mot strict?

List accepts anything and makes it into a list. Dict does the same. But they do it for purely technical reasons (but note it accepts everything that works). NumericProps accepts anything that is numeric like and doesn't convert. So why shouldn't bool? It probably should just convert it to a bool, although I don't really see the reason this is needed. But I cannot see why this should ever be enforced to only accept bools, that's not how python treats "truth".

**Comment by dessant** [comment]
@matham, I've always regarded the first part of some of the property names as a type hint, and that led me to follow the restriction I assumed, and because my assumtions were backed up by what I experienced with `StringProperty`.

I'm either extraordinarily confused about this issue as a core dev, or we can conclude that the regular Kivy user does expect type enforcement while using these properties. :stuck_out_tongue_winking_eye: As a side note, this was reported by a Kivy user on irc.

The error message is also clearly indicative of enforcing the boolean type:
https://github.com/kivy/kivy/blob/d437e93e8a769fc0a42cc1868f68feda8e4b1c17/kivy/properties.pyx#L918

But, if we are divided about the issue, we should offer both options (because it's cheap and aligns with user expectations) rather than pick one for each property, and let people remember which behavior is used with which property.

**Comment by matham** [comment]
Yes, but that `raise ValueError('%s.%s accept only bool' % ()` is never raised as far as I can see because I'm not sure if there's anything that doesn't inherit from object in py2. Or at least anything reasonable.

**Comment by dessant** [comment]
@matham, yes, but it does indicate the intentions of the author.

Anyway, that's not the main issue, but inconsistent behavior across properties.

**Comment by matham** [comment]
And you still have not given a example where type enforcement is useful. Also, although you might have expected that based on the name, I don't think that's how python typically operates. With the way things are weakly typed (again, you never should say `if val is True:...`).

To me sticking a option there is the lazy way out, unless there are actual reasons one would want both of the options. Because you're passing on the decision to the user when users never made that decision before. Or had reason to,

**Comment by dessant** [comment]
I'm not saying type enforcement is very useful in kivy's case, it's mostly just an user expectation (although it can help during development and debugging), but I do want to see consistent behavior across properties, either we try to coerce to the type indicated by the property name (and error if that's not possible), or we enforce the correct type.

**Comment by dessant** [comment]
We had a discussion with @matham, here are the main points:
* type hinting can be done by the user with the standard python api, to address the debugging benefit of type enforcement (without hard crashes)
* coercing a type is evil, but needed for some properties because of kivy internals, we should revisit if this type conversion can be eliminated in some cases
* if coercing is needed for the value, it should not be silent

@matham, it was not clear what happens with properties where we don't want coercion, because we are also against type enforcement, please clarify the foggy parts, maybe by listing the desired behavior for each typed property.

**Comment by dessant** [comment]
Also, here are some bugs we've found:
* NumericProperty converts `'5'` to `0.0` with a warning, while `'5.0'` is silently converted to `5.0`
* ListProperty silently converts strings to lists

**Comment by matham** [comment]
>  coercing a type is evil, but needed for some properties because of kivy internals, we should revisit if this type conversion can be eliminated in some cases

While not quite evil, I don't think they should be done, unless needed or justified because we should be doing duck typing. And so far I haven't seen a reason that it is justified. Only list/dict prop do this, and there there's no choice because we need to create the dispatchable list/dict. Numeric also does it but only when the input is e.g. a str so of course it's required then.

So by default it's not normally done (string is different, as I mentioned above about bytes/unicode).

> if coercing is needed for the value, it should not be silent

I think you meant `it should be silent`? StringProperty has some bugs where it emits a warning and gives a bad value. Those kinds of bugs should be fixed. But when it coerces (like when numeric converts str to float/int), I think it should be silent, unless there's an issue and then it should crash.

> @matham, it was not clear what happens with properties where we don't want coercion, because we are also against type enforcement, please clarify the foggy parts, maybe by listing the desired behavior for each typed property.

For bool, I think the fix is to make it dispatch only when the bool state changes, like @inclement said and not everytime when the value changes.

**Comment by dessant** [comment]
> I think you meant it should be silent?

I meant it should not be silent if it is not done the same way for all properties, but only a few.
Of course i don't mean a warning, but a debug log message when such conversion happens.

**Comment by dessant** [comment]
@matham, I do understand that this would lead to a million log messages for stuff like `NumericProperty('10dp')`, so maybe that leaves us with explicitly documenting type conversions instead?

**Comment by matham** [comment]
> @matham, I do understand that this would lead to a million log messages for stuff like NumericProperty('10dp'), so maybe that leaves us with explicitly documenting type conversions instead?

Yes. We just need to describe the type it accepts, e.g. str for numeric prop.

**Comment by akshayaurora** [comment]
For things that are not explicitly stated to be different, kivy follows python convention, after all we are a python framework.

- Forcing Boolean values to Boolean property...I agree with @matham here. We should follow python idioms here.

> For bool, I think the fix is to make it dispatch only when the bool state changes, like @inclement said and not overtime when the value changes.


+1 this seems like the correct fix for this.

**Comment by dessant** [comment]
Related issue for NumericProperty: https://github.com/kivy/kivy/pull/3392.
